ext/src/ruby_api/convert.rs

```diff
@@ -64,27 +64,32 @@ pub trait ToWasmVal {
 
 impl ToWasmVal for Value {
     fn to_wasm_val(&self, ty: ValType) -> Result<Val, Error> {
+        if ty.matches(&ValType::EXTERNREF) {
+            let extern_ref_value = match self.is_nil() {
+                true => None,
+                false => Some(ExternRef::new(ExternRefValue::from(*self))),
+            };
+
+            return Ok(Val::ExternRef(extern_ref_value));
+        }
+
+        if ty.matches(&ValType::FUNCREF) {
+            let func_ref_value = match self.is_nil() {
+                true => None,
+                false => Some(*<&Func>::try_convert(*self)?.inner()),
+            };
+            return Ok(Val::FuncRef(func_ref_value));
+        }
+
         match ty {
             ValType::I32 => Ok(i32::try_convert(*self)?.into()),
             ValType::I64 => Ok(i64::try_convert(*self)?.into()),
             ValType::F32 => Ok(f32::try_convert(*self)?.into()),
             ValType::F64 => Ok(f64::try_convert(*self)?.into()),
-            ValType::ExternRef => {
-                let extern_ref_value = match self.is_nil() {
-                    true => None,
-                    false => Some(ExternRef::new(ExternRefValue::from(*self))),
-                };
-
-                Ok(Val::ExternRef(extern_ref_value))
-            }
-            ValType::FuncRef => {
-                let func_ref_value = match self.is_nil() {
-                    true => None,
-                    false => Some(*<&Func>::try_convert(*self)?.inner()),
-                };
-                Ok(Val::FuncRef(func_ref_value))
-            }
             ValType::V128 => err!("converting from Ruby to v128 not supported"),
+            // TODO: to be filled in once typed function references and/or GC
+            // are enabled by default.
+            t => err!("unsupported type: {t:?}"),
```

Comment:
Nit / suggestion: I think it reads nicely when the RefType is a nested match:

```ruby
impl ToWasmVal for Value {
    fn to_wasm_val(&self, ty: ValType) -> Result<Val, Error> {
        match ty {
            ValType::I32 => Ok(i32::try_convert(*self)?.into()),
            ValType::I64 => Ok(i64::try_convert(*self)?.into()),
            ValType::F32 => Ok(f32::try_convert(*self)?.into()),
            ValType::F64 => Ok(f64::try_convert(*self)?.into()),
            ValType::V128 => err!("converting from Ruby to v128 not supported"),
            ValType::Ref(ty) => {
                if ty.matches(&RefType::EXTERNREF) {
                    let extern_ref_value = match self.is_nil() {
                        true => None,
                        false => Some(ExternRef::new(ExternRefValue::from(*self))),
                    };

                    Ok(Val::ExternRef(extern_ref_value))
                } else if ty.matches(&RefType::FUNCREF) {
                    let func_ref_value = match self.is_nil() {
                        true => None,
                        false => Some(*<&Func>::try_convert(*self)?.inner()),
                    };

                    Ok(Val::FuncRef(func_ref_value))
                } else {
                    // TODO: to be filled in once typed function references and/or GC
                    // are enabled by default.
                    err!("unsupported type: {ty:?}")
                }
            }
        }
    }
}
```

Same below. Use this version if you like it too, or disregard if you don't!